examples/mcpdft/03-metaGGA_functionals.py

```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env/python
+from pyscf import gto, scf, mcpdft
+
+mol = gto.M (
+    atom = 'O 0 0 0; O 0 0 1.2',
+    basis = 'ccpvdz',
+    spin = 2)
+
+mf = scf.RHF (mol).run ()
+
+# The translation of Meta-GGAs and hybrid-Meta-GGAs [ChemRxiv. 2024; doi:10.26434/chemrxiv-2024-5hc8g-v2]
+
+# Translated-Meta-GGA
+mc = mcpdft.CASCI(mf, 'tM06L', 6, 8).run ()
+
+# Hybrid-Translated-Meta-GGA
+tM06L0 = 't' + mcpdft.hyb('M06L',0.25, hyb_type='average')
+mc = mcpdft.CASCI(mf, tM06L0, 6, 8).run ()
+
+# MC23 on-top functional [ChemRxiv. 2024; doi:10.26434/chemrxiv-2024-5hc8g-v2]
+
+# MC23 = { '0.2952*HF + (1-0.2952)*rep-M06L, 0.2952*HF + (1-0.2952)*rep-M06L'}}
+# To calculate the MC23 energies,
+# First: Calculate the MCPDFT Energies using repM06L
+# Second: Add the CAS Contributation
+
+# State-Specific
+mc = mcpdft.CASCI(mf, 'trepM06L', 6, 8)
+e_pdft = mc.kernel()[0]
+emc23 = 0.2952*mc.e_mcscf + (1-0.2952)*e_pdft
```

Comment:
Ideally the end-user should not have to perform any "hybridization" themselves. What issue are you running into that you cannot set the `ot.hybrid` to `1-0.2952` when you register the new functional?

Reply:
are you not able to play around with [these variables](https://github.com/JangidBhavnesh/pyscf-forge/blob/e30dcc418ead44efe84d44149977bb4c067ba9af/pyscf/dft2/libxc.py#L63-L64)?

Reply:
Through some experimentation I figured out the basic recipe for this:

```
from pyscf import dft
from pyscf import dft2
dft.libxc = dft2.libxc
dft.numint.libxc = dft2.libxc
dft.numint.LibXCMixin.libxc = dft2.libxc

# Reparametrized-M06L: rep-M06L
# MC23 = { '0.2952*HF + (1-0.2952)*rep-M06L, 0.2952*HF + (1-0.2952)*rep-M06L'}}
# Here, I am registering the rep-M06L functional only.

# repM06L_C has 27 parameters
MC23_C =  np.array([0.06, 0.0031, 0.00515088, 0.00304966, 2.427648, 3.707473,
                    -7.943377, -2.521466, 2.658691, 2.932276, -8.832841e-01,
                    -1.895247, -2.899644, -5.068570e-01, -2.712838, 9.416102e-02,
                    -3.485860e-03, -5.811240e-04, 6.668814e-04, 0.0, 2.669169e-01,
                    -7.563289e-02, 7.036292e-02, 3.493904e-04, 6.360837e-04, 0.0, 1e-10])

# repM06L_X has 18 parameters
MC23_X =  np.array([3.352197, 6.332929e-01, -9.469553e-01, 2.030835e-01,
                    2.503819, 8.085354e-01, -3.619144, -5.572321e-01,
                    -4.506606, 9.614774e-01, 6.977048, -1.309337, -2.426371,
                    -7.896540e-03, 1.364510e-02, -1.714252e-06, -4.698672e-05, 0.0])

# See: pyscf/pyscf/dft/libxc_funcs.txt
XC_ID_MGGA_C_M06_L = 233
XC_ID_MGGA_X_M06_L = 203

libxc_register_code = 'mc23'.lower ()

dft2.libxc.register_custom_functional_(libxc_register_code, 'M06L',
                                    ext_params={XC_ID_MGGA_X_M06_L: MC23_X,
                                                XC_ID_MGGA_C_M06_L: MC23_C},
                                    hyb=(0.2952,0,0),facs=(0.7048,0.7048))

mc = mcpdft.CASCI (mf, 'tmc23', 6, 8).run ()
```

There is some extra housekeeping you have to do to make this work:
1. switch `mcpdft._libxc` from `dft.libxc` to `dft2.libxc`
2. remove an obvious `NotImplementedError` in `mcpdft.otfnal`
3. make some logic branch or try/except block in `otfnal._hybrid_2c_coeff`

In addition, one probably ought to also
1. Run all the tests, including the usually-skipped `_slow` tests, to make sure none of this broke anything, and
2. Define `'MC23'` in `otfnal.OT_HYB_ALIAS` so that people can use the actual name of the functional in the standard way.l

Reply:
@Dayou-Zhang the behavior of `dft2.libxc` is a little strange in the code snipped I posted above because if I define `mc23` off of M06 instead of M06-L, placing the hybridization parameter at the appropriate position in the appropriate list, it ends up _adding_ to the hybridization parameter I entered in the `hyb` kwarg and I end up with  overall hyb=0.5904. I decided not to do this because as I said above, hybridization just fundamentally works differently between KS-DFT and MC-PDFT, but I still think this behavior is odd and I don't see in the code how it comes about. Any comments?

Reply:
Thanks. 
I will add this.

Reply:
> @Dayou-Zhang the behavior of `dft2.libxc` is a little strange in the code snipped I posted above because if I define `mc23` off of M06 instead of M06-L, placing the hybridization parameter at the appropriate position in the appropriate list, it ends up _adding_ to the hybridization parameter I entered in the `hyb` kwarg and I end up with overall hyb=0.5904. I decided not to do this because hybridization just fundamentally works differently between KS-DFT and MC-PDFT, but I still think this behavior is odd and I don't see in the code how it comes about. Any comments?

Returning hyb=0.5904 is in fact the expected behavior. The `hyb` kwarg overwrites the `hyb` variable returned by `parse_xc` instead of directly setting the total HF exchange. So `hyb[0]` actually represents additional HF exchange that are not part of the LibXC functional components. This can be confirmed by running
```
from pyscf import dft
print(dft.libxc.parse_xc('M06'))
```
which prints
```
((0, 0, 0), ((np.int32(449), 1), (np.int32(235), 1)))
```
We can see `hyb[0]` is zero, not 0.27.

I now realize that the `hyb` kwarg is really counterintuitive. I will try to re-design the API to make it behave more intuitively. Feel free to provide any suggestions if you have any.

Reply:
I am afraid that supporting strings like `'0.75*mypurefunc+0.25*HF'` is kind of challenging, if at all possible. I would call this "recursively defined functionals" since they are constructed from a previously customized functional.

There are two methods I could think of to make recursively defined functionals possible. In the first method, we do not change the Python API functions such as `eval_xc`. Since all C API functions take an array of `xc_func_type` structs, we need to make a new `xc_func_type` array for the recursively defined functional that copies the `xc_func_type` structs from the previously defined functional component and update coefficients like `hyb` and `fn_facs`. However, LibXC currently does not have an API that performs a deep copy of the functional data structure. We could do a shallow copy of `xc_func_type` with `memcpy`, but this would be problematic if users attempt to set functional parameters on the recursively defined functionals, as it may also change the custom functional that was defined first. Plus this is not memory safe and can cause segfault if the original functional is destructed. In addition, having two functional components with the same integer ID is now possible in recursively defined functionals. The two functional components are likely to be not identical since users could have performed customization on one of them, so we cannot simply merge them by the same integer ID. This would cause the `ext_params` kwarg to fail, since the integer ID is no longer a unique identifier of the functional component.

Alternatively, we can rewrite all the python API functions so that they will call the API functions of the functional components recursively. This would require special handling of the custom functional components separately from the pure LibXC functional components. In this way, we don't need to worry about the two problems in the first method, but it will certainly be much more work. It would be less efficient because of the recursive API call and may also slightly impact the performance of functionals that are not recursively defined.

Because of the complications, I am hesitating whether we should support recursively defined functionals. Since all the functionals are linear combinations, one should probably be able to define any recursively defined functionals in a non-recursive way.

Reply:
I'm not sure I agree, but we're wandering off topic from Bhavnesh's PR so lets leave it there for now.